Day/Night Cycle Improvments

randomPoison opened the pull request:
- Added a fade in in the morning so that you don't see the player teleport.
- Added a fixed portion of the day/night cycle that stays night at the end.
- Made night settings for all renders darker.
- Made the skybox texture darker.
- Made the fade out effect last a fixed amount of time.

randomPoison committed:
Longer Day/Night Cycle
fixes #65
- Increased the length of day/night cycle.
- Added a fade in effect for the morning.
- Reduced the time of the fade out at midnight.

randomPoison committed:
Longer Night Time
- Make night fall earlier and stay dark longer.
- Made the different render zones into prefabs.
- Added a debug option to DayCycleManager to disable the midnight reset.

randomPoison committed:
Darker Night Time
- Changed the skybox texture to be darker.
- Made night colors darker.

randomPoison committed:
Improved Day Cycle Manager
- Removed constant checks to perform day/night transitions and replaced them with cleaner Invoke() calls and coroutines.

brickman1444 commented on Assets/Scripts/Player/PlayerControls.cs:
```diff
@@ -58,7 +58,7 @@ void Awake()
 
-	void Update()
+	void FixedUpdate()
 	{
```
Why was this change made? Isn't input usually handled on Update() ?

randomPoison replied:
This fixes the problem of one click giving multiple items. It might also help with some of the jumping problems, but I'm not sure. When we were reading input in `Update()` input was being read less often than physics was being updated in `FixedUpdate()` so 2 or 3 frames would be using the same input data. This fixes it for now, though we should seek out a better solution at some point.

brickman1444 commented on Assets/Scripts/Utilities/DayCycleManager.cs:
```diff
@@ -4,55 +4,61 @@
+
 	[Tooltip( "The length (in seconds) of a day" )]
-	[SerializeField] float _dayCycleLength = 60.0f;
+	[SerializeField] float _dayCycleLength = 300.0f;
```
You shouldn't be worried about updating this when you change the value in the inspector. Just keep it to a sane default value.

randomPoison replied:
I was trying to change it to a sane default value.

brickman1444 commented on Assets/Scripts/Utilities/DayCycleManager.cs:
```diff
@@ -61,34 +67,64 @@ public static void RegisterEndOfDayCallback( EndOfDayCallback callback )
-		Color overlayColor = _midnightOverlay.color;
-		overlayColor.a = 0.0f;
-		_midnightOverlay.color = overlayColor;
+		StartCoroutine( FadeOutMidnightOverlay() );
```
Are StartMidnightOverlay() and StartMorningOverlay() worth having? They both only have one line each. I think it'd be clearer to remove them

randomPoison replied:
`StartMidnightOverlay()` is needed because it's called by an `Invoke()` (blech), but I can get rid of `StartMorningOverlay()`.

brickman1444 commented on Assets/Scripts/Utilities/DayCycleManager.cs:
```diff
@@ -61,34 +67,64 @@ public static void RegisterEndOfDayCallback( EndOfDayCallback callback )
+		{
+			if ( elapsedTime < _blackoutDuration )
+			{
+				elapsedTime += Time.deltaTime;
```
Do you want to be ticking inside this check?

randomPoison replied:
It's functionally the same as ticking before the check, and I think this is a bit cleaner.

brickman1444 commented on Assets/Scripts/Utilities/DayCycleManager.cs:
```diff
@@ -61,34 +67,64 @@ public static void RegisterEndOfDayCallback( EndOfDayCallback callback )
+				elapsedTime += Time.deltaTime;
+				Color overlayColor = _midnightOverlay.color;
+				overlayColor.a = elapsedTime / _blackoutDuration;
+				_midnightOverlay.color = overlayColor;
```
If the color can't be changed directly you should go ahead and write the extension methods so we can do:

_midnightOverlay.color = _midnightOverlay.color.SetA( elapsedTime / _blackoutDuration );

brickman1444 commented on Assets/Scripts/Utilities/DayCycleManager.cs:
```diff
@@ -61,34 +67,64 @@ public static void RegisterEndOfDayCallback( EndOfDayCallback callback )
+		float elapsedTime = 0.0f;
+		while ( true )
+		{
+			if ( elapsedTime < _blackoutDuration )
```
Consider reformatting to 

```
elapsedTime += Time.deltaTime;
Color overlayColor = _midnightOverlay.color;
overlayColor.a = elapsedTime / _blackoutDuration;
_midnightOverlay.color = overlayColor;

if ( elapsedTime >= _blackoutDuration )
{
    yield return break;
}

yield return null;
```

brickman1444 commented on Assets/Scripts/Utilities/RenderSettingsManager.cs:
```diff
@@ -3,23 +3,23 @@
+	public Color lightColor;
+	public float lightIntensity;
+	public Color fogColor;
+	public float fogDensity;
```
How does Unity fill these in when you create a new one in the inspector? 0's and whites?

randomPoison replied:
Just 0's. So the default's become less reasonable, but we produce less garbage on updates. I think it's worthwhile.

brickman1444 commented on Assets/Scripts/Utilities/RenderSettingsManager.cs:
```diff
@@ -96,7 +99,16 @@ void Awake()
 	void Update()
 	{
-		daylightIntensity = Mathf.Cos( DayCycleManager.dayCycleTimer / DayCycleManager.dayCycleLength * 2.0f * Mathf.PI ) * -0.5f + 0.5f;
+		float timeOfDay = DayCycleManager.currentTime / DayCycleManager.dayCycleLength;
```
I would put the night fraction adjustment in here.

```
timeOfDay = DayCycleManager.currentTime / ( DayCycleManager.dayCycleLength * ( 1.0f - _nightFraction) );
```

Then you compare it to 1.0f in the if statement and take out the division from the cosine line

randomPoison replied:
I can do you one better:

```
float timeOfDay = timeOfDay = DayCycleManager.currentTime / ( DayCycleManager.dayCycleLength * ( 1.0f - _nightFraction) );
timeOfDay = Mathf.Clamp01( timeOfDay );
daylightIntensity = Mathf.Cos( timeOfDay * 2.0f * Mathf.PI ) * -0.5f + 0.5f;
```

brickman1444 commented on Assets/Scripts/Utilities/RenderSettingsManager.cs:
```diff
@@ -77,12 +84,8 @@ TimeRenderSettings _currentTimeRenderSettingsProperty
 	[ReadOnly("Daylight Intensity"), Tooltip( "0 is midnight. 1 is noon." )]
-	[SerializeField]
-	float _daylightIntensity = 1.0f;
+	[SerializeField] float _daylightIntensity = 1.0f;
```
Why does this serialize? You never would really want to set it. It would be changed again the next frame.

randomPoison replied:
It has to be marked as `[SerializeField]` or `public` for `[ReadOnly]` to work. We should probably see if we can change `[ReadOnly]` to allow non-serialized members to still show up.

brickman1444 commented on Assets/Scripts/Utilities/RenderSettingsManager.cs:
```diff
@@ -3,23 +3,23 @@
+	public Color fogColor;
+	public float fogDensity;
 
 	public static TimeRenderSettings Lerp(TimeRenderSettings fromSettings, TimeRenderSettings toSettings, float t)
```
consider adding lerp methods to these structs that take in the output location by reference. It would save on copying memory around.

brickman1444 commented:
done

randomPoison committed:
Code Review Changes
- Removed redundant StartMorningOverlay() method.
- Updated Lerp methods for TimeRenderSettings and RenderSettingsData to cut down and copying data.
- Streamlined the time of day calculation.
- Had to remove _currentTimeRenderSettingsProperty and split its functionality into UpdateRenderSettings().

randomPoison committed:
Merge remote-tracking branch 'origin/master' into game_feel

Conflicts:
	Assets/Prefabs/Actors/Player.prefab
	Assets/Scenes/BlockOut.unity
